File: solutions/053-maximum-subarray/solution.py
from typing import List


class Solution:
    def max_sub_array(self, nums: List[int]) -> int:
        max_sum = nums[0]
        for i in range(1, len(nums)):
            if nums[i - 1] > 0:
                nums[i] += nums[i - 1]
            if nums[i] > max_sum:
                max_sum = nums[i]
        return max_sum


if __name__ == "__main__":
    sol = Solution()
    got = sol.max_sub_array([-2, 1, -3, 4, -1, 2, 1, -5, 4])
    status = "PASS" if got == 6 else "FAIL"
    print(f"{status} | sum={got} (expected 6)")

File: solutions/053-maximum-subarray/solution.go
package main

import "fmt"

func maxSubArray(nums []int) int { //思路，只要前面的数字和大于零，就把他加上，如果小于0，则当前就是最大的，重新计算

	n := len(nums)
	maxSum := nums[0]
	for i := 1; i < n; i++ {
		if nums[i-1] > 0 {
			nums[i] += nums[i-1]
		}
		if nums[i] > maxSum {
			maxSum = nums[i]
		}
	}
	return maxSum
}

func main() {
	got := maxSubArray([]int{-2, 1, -3, 4, -1, 2, 1, -5, 4})
	fmt.Printf("PASS | sum=%d\n", got)
}
